File: src/www/__test__/testSteps/AbstractPostLineStep.ts
import { PostLineStep } from "../testPost/PostLineStep";
import { PostLine } from "../testPost/reader";

export class AbstractPostLineStep implements PostLineStep {
  constructor(private regexp: RegExp) {}

  match(line: PostLine): boolean {
    return !!line.match(this.regexp);
  }

  async run(line: PostLine) {
    var match = line.match(this.regexp);
    await this.runMatch(line, match || []);
  }

  async runMatch(line: PostLine, match: string[]) {
    throw new Error("Cannot call an abstract method");
  }

  getPrettyPrint(): string {
    return `Matches "${this.regexp}" (${this.constructor.name})`;
  }
}

File: src/test/java/com/drpicox/game/testSteps/AbstractPostLineStep.java
package com.drpicox.game.testSteps;

import com.drpicox.game.testPost.reader.PostLine;
import com.drpicox.game.testPost.PostLineStep;

public abstract class AbstractPostLineStep implements PostLineStep {
    private final String regex;

    public AbstractPostLineStep() {
        this.regex = getRegex();
    }

    @Override
    public boolean match(PostLine line) {
        return line.match(regex) != null;
    }

    @Override
    public void run(PostLine line) {
        var match = line.match(regex);
        run(line, match);
    }

    protected abstract String getRegex();
    protected abstract void run(PostLine line, String[] match);

    @Override
    public String getPrettyPrint() {
        return "Matches \"" + regex + "\" (" + this.getClass().getName() + ")";
    }
}
